README.md
# WeeWX-BCRobotics V3
A Python V3 WeeWX driver for the BC Robotics and SparkFun weather station Raspberry Pi based hardware.

weewx-bcrobo

Driver to collect data from the "Spark Fun" SEN-08942 RoHS weather meters 
connected to the BC Robotics interface.

See:
https://www.sparkfun.com/products/8942

https://www.bc-robotics.com/tutorials/raspberry-pi-weather-station-part-1/

INTRODUCTION

The BC Robotics interface uses the Aidafruit BME280 sensor for humidity, pressure, 
and temperature along with a ADS1x15 ADC (analog to digital converter) to read the 
wind direction. The rain and wind speed are measured through the GPIO interface by 
counting "ticks" (rotations or measures of water). The driver includes a erroneous
rain tick filter, since this sensor can emit single undesired ticks.

If you want to build your own system see the documentation provided in the attached "Issue".

SOFTWARE SETUP

Before installing the BC Robotics driver, setup the required software by running the 
following commands (in this order):

[Install the Adafruit Python GPIO Library]

sudo apt-get update
sudo pip3 install --upgrade setuptools
pip3 install RPI.GPIO
pip3 install adafruit-blinka
  

[Install the Adafruit BME280 Library]

sudo pip3 install adafruit-circuitpython-bme280

[Install the ADS1x15 Library, the ADS1015 Analog to Digital chip (wind direction)]

sudo pip3 install adafruit-circuitpython-ads1x15  

[Setup the DS18B20 temperature sensor and install the software library]

cd
sudo modprobe w1-gpio
sudo modprobe w1-therm
cd /sys/bus/w1/devices
ls 

  
cd
sudo apt-get install python-w1thermsensor
  
  
The “ls” command will display the contents of the devices 'folder' in the window. The 
DS18B20 shows up as an address something like "28-0316853d8fff" – but each sensor has 
a unique ID.

TESTING THE HARDWARE

You can test the sensors by running the included test app: BCRobotics-test-app.py

Run the Python IDE:

  >>thonny

Now open the "BCRobotics-test-app.py" test app from the IDE and hit 'F5' to run it. It will 
continuously print out the readings from the sensors, including the value read from the ADC 
for the wind direction. Use these values (and a test template for the wind direction) to 
double check the values used in the driver. Hit "<ctrl> c" to stop the program.


DRIVER INSTALLATION

1) install weewx (see the weewx user guide)

2) download the driver and the documentation:

wget -O BCRobotics.zip https://github.com/David-Enst/WeeWX-BCRobotics/archive/master.zip

3) Extract the BCRobotics.py driver file and copy it to the Pi here:

usr/share/weewx/user/BCRobotics.py

4) Edit the etc/weewx/weewx.config file as follows:

Change the station type:
station_type = BCRobotics

Add the BCRobotics driver definition:
[BCRobotics]
    # This defines the "Spark Fun" SEN-08942 / BC Robotics weather stations.
    # See:    https://www.sparkfun.com/products/8942
    #         https://www.bc-robotics.com/tutorials/raspberry-pi-weather-station-part-1/
    
    # The time (in seconds) between LOOP packets.
    loop_interval = 3
    
    # Driver mode - tcp, udp, or serial
    mode = serial
    
    # If serial, specify the serial port device. (ex. /dev/ttyS0, /dev/ttyUSB0,
    # or /dev/cuaU0)
    # If TCP, specify the IP address and port number. (ex. 192.168.36.25:3000)
    port = /dev/ttyS0
    
    # The amount of time, in seconds, before the connection fails if there is
    # no response
    timeout = 3
    
    # Debug level - the level of message logging. The higher
    # the number, the more info is logged.
    debug_read = 0
    
    # The driver to use:
    driver = user.BCRobotics
    
Verify the definition of [[[Units]]] as follows:
        [[[Units]]]
            [[[[Groups]]]]
                group_altitude    = meter
                group_pressure    = hPa
                group_rain        = mm
                group_rainrate    = mm_per_hour
                group_temperature = degree_C
                group_degree_day  = degree_C_day
                group_speed       = km_per_hour
                group_speed2      = km_per_hour2

Verify the definition of [[[Labels]]] to ensure the web pages generated display the correct labels:
        [[[Labels]]]
            # Generic labels, keyed by an observation type.
            [[[[Generic]]]]
                barometer      = Barometer
                dewpoint       = Dew Point
                ET             = ET
                heatindex      = Heat Index
                inHumidity     = Case Humidity
                inTemp         = Case Temperature
                
Update the [StdConvert] section is to reflect the fact that all data in the database is stored in METRIC by default:
    target_unit = METRIC    # Options are 'US', 'METRICWX', or 'METRIC'
    
The database and driver is setup to store values in METRIC by default. Therefore, the following updates to the quality control parameters are required:
[StdQC]
    [[MinMax]]
        barometer =   800, 1110, mbar # = hPa
        pressure =    800, 1110, mbar  # = hPa
        outTemp =     -50, 49, degree_C
        inTemp =      -25, 49, degree_C
        outHumidity = 0, 100
        inHumidity =  0, 100
        windSpeed =   0, 300, km_per_hour
        rain =        0, 250, mm

Finally, update the archive section to force WeeWX to generate archive records through software generation, since the BCRobotics hardware does not generate them:
[StdArchive]
    # If possible, new archive records are downloaded from the station
    # hardware. If the hardware does not support this, then new archive
    # records will be generated in software.
    # Set the following to "software" to force software record generation.
    record_generation = software
   
    
5) Restart WeeWX:

sudo /etc/init.d/weewx stop
sudo /etc/init.d/weewx start

 NOTE: We have moved to the 21st Century, so the database is in METRIC!
       Therefore, if you have a database created already, then see:

         http://weewx.com/docs/customizing.htm#Changing_the_unit_system

CREDITS

Implementation of this driver was much easier thanks to the work by others, 
including BC-Robotics:

Joerg Raedler
  https://bitbucket.org/jraedler/sunnywebbox

Aleksandar Tsankov
  https://github.com/fullergalway/weewx-airmar-150wx-driver

Matthew Wall    
  https://github.com/matthewwall/weewx-sdr


changelog
0.0.2 07 Jan 2019
* fix initialization

0.0.3 09 Jan 2019
* fix GPIO event handling

1.0.4 09 Jan 2019
* first release, fix logging error

1.0.6 09 Jan 2019
* fix ADC / windDir error

1.0.7 14 Jan 2019
* fix rain error

1.0.12 15 Jan 2019
* adjust humidity for outTemp vs. case_temp 

1.0.16 17 Jan 2019
* catch single erronious rain ticks 

1.0.17 19 Jan 2019
* fix wind direction error 

1.0.24 6 Feb 2019
* improved humidity calculation
 
2.0.27 26 Aug 2020
* Updated the driver to Python V3


BCRobotics.py
# Copyright 2019 David W. Enstrom, all rights reserved
"""
Driver to collect data from the "Spark Fun" SEN-08942 RoHS
weather meters and the BC Robotics interface.

See:    https://www.sparkfun.com/products/8942
        https://www.bc-robotics.com/tutorials/raspberry-pi-weather-station-part-1/

"""

import syslog
import threading
import time
import math

from w1thermsensor import W1ThermSensor
from Adafruit_BME280 import *
import Adafruit_ADS1x15
import RPi.GPIO as GPIO

import weewx.drivers
import weewx.units
import weewx.accum

DRIVER_NAME = 'BCRobotics'
DRIVER_VERSION = '1.0.17'

windTick = 0     # Count of the wind speed input trigger
rainTick = 0     # Count of the rain input trigger
interval = 3     # Set now to define scope of variable
rainTime = 0     # Use this to detect erroneous rain events

try:
     ds18b20 = W1ThermSensor()
except Exception as err:
     loginf('Error setting up Temperature Sensor %s' % err)
     TempSensor = False
else:
     TempSensor = True

bme = BME280(t_mode=BME280_OSAMPLE_8, p_mode=BME280_OSAMPLE_8, h_mode=BME280_OSAMPLE_8)
adc = Adafruit_ADS1x15.ADS1115()



def loader(config_dict, _):
    return BCRoboDriver(**config_dict[DRIVER_NAME])

def confeditor_loader():
    return BCRoboConfEditor()

class BCRoboConfEditor(weewx.drivers.AbstractConfEditor):
    @property
    def default_stanza(self):
        return """

[BCRobotics]
    # This section is for the "Spark Fun" SEN-08942 / BC Robotics weather stations.
    # See: https://www.sparkfun.com/products/8942
    #      https://www.bc-robotics.com/tutorials/raspberry-pi-weather-station-part-1/


    # The time (in seconds) between LOOP packets. Default is:
    loop_interval = 3

    # Driver mode - tcp, udp, or serial
    mode = serial

    # If serial, specify the serial port device. (ex. /dev/ttyS0, /dev/ttyUSB0,
    # or /dev/cuaU0)
    # If TCP, specify the IP address and port number. (ex. 192.168.36.25:3000)
    port = /dev/ttyS0

    # The amount of time, in seconds, before the connection fails if there is
    # no response
    timeout = 3

    # Debug level - the level of message logging. The higher
    # the number, the more info is logged.
    debug_read = 0

    # The driver to use:
    driver = user.BCRobotics
    
"""

# flags for enabling/disabling debug verbosity
DEBUG_COMM = 0
DEBUG_CONFIG_DATA = 0
DEBUG_WEATHER_DATA = 0
DEBUG_HISTORY_DATA = 0
DEBUG_DUMP_FORMAT = 'auto'


def logmsg(dst, msg):
    syslog.syslog(dst, 'BCRobo: %s: %s' %
                  (threading.currentThread().getName(), msg))
    
def logdbg(msg):
    logmsg(syslog.LOG_DEBUG, msg)


def loginf(msg):
    logmsg(syslog.LOG_INFO, msg)


def logcrt(msg):
    logmsg(syslog.LOG_CRIT, msg)


def logerr(msg):
    logmsg(syslog.LOG_ERR, msg)


def logtee(msg):
    loginf(msg)
    print "%s\r" % msg


class BCRoboDriver(weewx.drivers.AbstractDevice):
    """weewx driver that communicates with a "Spark Fun" SEN-08942 / BC Robotics station

    mode - Communication mode - TCP, UDP, or Serial.
    [Required. Default is serial]

    
    port - Serial port or network address.
    [Required. Default is /dev/ttyS0 for serial,
     and 192.168.36.25:3000 for TCP/IP]

    loop_interval - The time (in seconds) between LOOP packets.
    [Required. Default is loop_interval = 3.0]
    
    timeout - The amount of time, in seconds, before the connection fails if
    there is no response.
    [Optional. Default is 3]

    debug_read - The level of message logging. The higher this number, the more
    information is logged.
    [Optional. Default is 0]
    
    """
    def __init__(self, **stn_dict):
         
         loginf('Driver version is %s' % DRIVER_VERSION)
         global interval
         interval = int(stn_dict.get('loop_interval', 3))
         #loginf('Interval is %s' % interval)
         global rainTime
         rainTime = int(time.time())   # used to detect erronious rain ticks

         try:
              # Set GPIO pins to use BCM pin numbers
              GPIO.setmode(GPIO.BCM)
              
              # Set digital pin 17 to an input and enable the pullup (wind speed)
              GPIO.setup(17, GPIO.IN, pull_up_down=GPIO.PUD_UP)

              # Set digital pin 23 to an input and enable the pullup (rain)
              GPIO.setup(23, GPIO.IN, pull_up_down=GPIO.PUD_UP)

              # Define event to detect wind (4 ticks per revolution)
              #  1 tick/sec = 1.492 mph or 2.4 km/h
              #  So: wind = (windTick * 2.4) / loop_interval
              GPIO.add_event_detect(17, GPIO.BOTH)
              def windtrig(self):
                  global windTick
                  windTick += 1
                
              GPIO.add_event_callback(17, windtrig)

              # Define event to detect rain (0.2794mm per tick)
              # with a 1hr time period where the first tick is ignored
              # which is used to detect an improper single tick
              GPIO.add_event_detect(23, GPIO.FALLING)
              def raintrig(self):
                  global rainTick
                  global rainTime
                  if int(time.time())-rainTime >= 3600:
                      loginf('Possible bad rain tick')
                      rainTime = int(time.time())
                  else:
                      rainTick += 1
                      rainTime = int(time.time())
              
              GPIO.add_event_callback(23, raintrig)
              
         except Exception as err:
               loginf('Error setting up GPIO: ' % err)
         else:
               loginf('GPIO setup fine.')




    @property
    def hardware_name(self):
        return "BCRobotics"
    
    def genLoopPackets(self):
        while True:
            packet = {'dateTime': int(time.time() + 0.0),
                      'usUnits': weewx.METRIC}
            readings = StationData.get_readings()
            packet.update(readings)
    #        self._augment_packet(packet)
            yield packet

    #def _augment_packet(self, packet):
    #    # calculate the rain delta from rain total
    #    if self.last_rain is not None:
    #        packet['rain'] = packet['rain_total'] - self.last_rain
    #    else:
    #        packet['rain'] = None
    #    self.last_rain = packet['rain_total']

# ======================================================================= #
#       StationData class - setup stn and get data from the device        #
# ======================================================================= #


class StationData():
    def __init__(self):
       # Initialize the setup and define ports
       last_rain = None
       global windTick
       global rainTick
       windTick = 0   # Count of the wind speed input trigger
       rainTick = 0   # Count of the rain input trigger



    @staticmethod
    def get_readings():
        """ The "Spark Fun" SEN-08942 / BC Robotics station emits data
         through a BME280 interface, a ds18b20 temperature sensor plus
         a ADS1015 Analog to Digital chip for the wind direction.

        https://www.bc-robotics.com/tutorials/raspberry-pi-weather-station-part-2/

        The following data is provided by the sensor hardware:

          windSpeed   - from total ticks (1 tick/sec = 2.4 km/h)
          windDir     - wind direction (0-255)
          outTemp     - outdoor temperature (in Deg C)
          rain        - rain during the loop_interval (1 tick = 0.2794 mm)
          pressure    - pressure (in Pascals)
          inTemp      - hardware case temperature (in Deg C)
          outHumidity - outdoor humidity (+/-0.1 %)
          
        """
        #
        # Get the readings, define the data list, and fill it in from readings
        # NOTE: The pressure reading is 'pressure', not 'barometer'
        #

        #if global TempSensor
        # Get temperature from DS18B20 sensor in degrees_C
        outTemp = ds18b20.get_temperature()
        
        # Get Temperature from BME280 in degrees_C
        case_temp = bme.read_temperature()
        
        # Get Barometric Pressure from BME280 in Pascals
        #   and convert to mbar
        #   100 pascal = 1 mbar, or 0.00029529983071445 inhg
        pressure = bme.read_pressure() / 100
        
        # Get Humidity (inside the case) from BME280 in %
        humidity = bme.read_humidity()
        
        # This humidity is measured inside the case, 
        # which is warmer than the ambient air. Therefore
        # it is converted to external humitity based upon  
        # the case_temp and outTemp. First calculate the 
        # absolute moisture. 
        absMoisture = humidity * 0.42 * math.exp(case_temp * 0.06235398)/10
        
        # Adjust humidity reading to the outside temperature
        humidity = absMoisture * 10 / (0.42 * math.exp(outTemp * 0.06235398))

        # Calculate wind direction (angle) based on ADC reading
        #   Read ADC channel 0 with a gain of 1
        windDir = 1.5

        val = adc.read_adc(0, gain=1)

        if 19600 <= val <= 20999:
            windDir = 0

        if 9000 <= val <= 10799:
            windDir = 22.5

        if 10800 <= val <= 13999:
            windDir = 45

        if 2000 <= val <= 2299:
            windDir = 67.5

        if 2300 <= val <= 2999:
            windDir = 90

        if 1000 <= val <= 1999:
            windDir = 112.5

        if 4000 <= val <= 4999:
            windDir = 135

        if 3000 <= val <= 3999:
            windDir = 157.5

        if 6600 <= val <= 8999:
            windDir = 180

        if 5000 <= val <= 6599:
            windDir = 202.5

        if 15900 <= val <= 16999:
            windDir = 225

        if 14000 <= val <= 15899:
            windDir = 247.5

        if 24000 <= val <= 24999:
            windDir = 270

        if 21000 <= val <= 21999:
            windDir = 292.5

        if 22000 <= val <= 23999:
            windDir = 315

        if 17500 <= val <= 19599:
            windDir = 337.5

        if windDir == 1.5:
            value = str(val)
            loginf('Wind direction error: ' + value)


        # Calculate the average wind speed over this 'interval' (km/h)
        #  1 tick/sec = 1.492 mph or 2.4 km/h
        #  So: (windTick * 2.4) / loop_interval
        global windTick
        windSpeed = (windTick * 2.4) / interval
        windTick = 0

        # Calculate the rainfall over this 'interval' (cm)
        # 1 tick = 0.011 inches or 0.02794 cm 
        global rainTick
        rain = rainTick * 0.02794
        
        if rain > 0:
            raintxt = str(rain)
            loginf('Rain is falling: ' + raintxt)
            rainRate = (rain / interval) * 3600  # cm/h
            rainTick = 0
        else:
            rainRate = None
            rain = None 
            
        data = dict()
        data['windSpeed'] = windSpeed   # km/h
        data['windDir'] = windDir       # compass deg
        data['outTemp'] = outTemp       # degree_C
        data['rain'] = rain             # cm as per default
        data['rainRate'] = rainRate     # cm/hr
        data['pressure'] = pressure     # mbar
        data['inTemp'] = case_temp      # degree_C
        data['outHumidity'] = humidity  # percent
        
        rain = 0
        rainRate = 0
        return data

    def __exit__(self, _, value, traceback):
        #self.close()
        return 

    def __enter__(self):
        #self.open()
        return self


# Define a main entry point for basic testing of the station without weewx
# engine and service overhead.  Invoke this as follows from the WeeWX root dir (i.e., /usr/bin/weewxd):
#
# sudo PYTHONPATH=/usr/share/weewx python /usr/share/weewx/user/BCRobotics.py
# 

if __name__ == '__main__':

    syslog.openlog('BCRobotics', syslog.LOG_PID | syslog.LOG_CONS)
    syslog.setlogmask(syslog.LOG_UPTO(syslog.LOG_DEBUG))
    
    with StationData() as s:
        #s.set_logger_mode()

        time.sleep(interval)
        while True:
            print time.time(), s.get_readings()




BCRobotics-test-app.py
import time
import time
import board
import busio
import adafruit_bme280

from w1thermsensor import W1ThermSensor
import adafruit_ads1x15.ads1015 as ADS
from adafruit_ads1x15.analog_in import AnalogIn

import RPi.GPIO as GPIO


out_temp = 0

try:
    ds18b20 = W1ThermSensor()
    print ('DS18b20 temp sensor OK')
    TempSensor = True
except Exception as err:
    print ('Temperature Sensor Error:', err)
    TempSensor = False

# Create library object using our Bus I2C port
i2c = busio.I2C(board.SCL, board.SDA)

try:
     bme = adafruit_bme280.Adafruit_BME280_I2C(i2c) # temp, pressure, humidity
     PressSensor = True
except Exception as err:
     print ('BME280 Pressure Error:', err)
     PressSensor = False
 
try:
     ads = ADS.ADS1015(i2c)        # wind and rain
     ads.gain = 1
     chan = AnalogIn(ads, ADS.P0)
     WindSensor = True
except Exception as err:
     print ('Wind Sensor Error:', err)
     WindSensor = False

bme.sea_level_pressure = 1013.25 # generic value
interval = 3   # Time between loops (seconds)
windTick = 0   # Count of the wind speed input trigger
rainTick = 0   # Count of the rain input trigger
val      = 0   # Wind direction

# Set GPIO pins to use BCM pin numbers
GPIO.setmode(GPIO.BCM)

# Set digital pin 17 to an input and enable the pullup (wind speed)
GPIO.setup(17, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# Set digital pin 23 to an input and enable the pullup (rain)
GPIO.setup(23, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# Event to detect wind (4 ticks per revolution)
GPIO.add_event_detect(17, GPIO.BOTH)
def windtrig(self):
    global windTick
    windTick += 1

GPIO.add_event_callback(17, windtrig)

# Event to detect rain (0.2794mm per tick)
GPIO.add_event_detect(23, GPIO.FALLING)
def raintrig(self):
    global rainTick
    rainTick += 1

GPIO.add_event_callback(23, raintrig)

while True:
   
    # Get temperature from DS18B20 sensor
    if TempSensor : out_temp = ds18b20.temperature()

    # Get Temperature, humidity & Pressure from BME280
    if PressSensor : 
        case_temp = bme.temperature 
        pressure_pa= bme.pressure
        #pressure_pa = pressure_pa + 87/8.3 # adjust to sea level
        pressure = pressure_pa    # leave in hPa = mBar
        humidity = bme.humidity

    # Calculate wind direction based on ADC reading
    #   Read ADC channel 0 with a gain of 1
    if WindSensor : val = chan.value

    windDir = "Not connected"

    if 19600 <= val <= 20999:
        windDir = "N"

    if 9000 <= val <= 10799:
        windDir = "NNE"

    if 10800 <= val <= 13999:
        windDir = "NE"

    if 2000 <= val <= 2299:
        windDir = "ENE"

    if 2300 <= val <= 2999:
        windDir = "E"

    if 1000 <= val <= 1999:
        windDir = "ESE"

    if 4000 <= val <= 4999:
        windDir = "SE"

    if 3000 <= val <= 3999:
        windDir = "SSE"

    if 6600 <= val <= 8999:
        windDir = "S"

    if 5000 <= val <= 6599:
        windDir = "SSW"

    if 15900 <= val <= 17499:
        windDir = "SW"

    if 14000 <= val <= 15899:
        windDir = "WSW"

    if 24000 <= val <= 24999:
        windDir = "W"

    if 21000 <= val <= 21999:
        windDir = "WNW"

    if 22000 <= val <= 23999:
        windDir = "NW"

    if 17500 <= val <= 19599:
        windDir = "NNW"

    # Calculate the average wind speed over 
    #   this 'interval' in km/h
    windSpeed = (windTick * 1.2) / interval
    windTick = 0

    #Calculate the rainfall over this 'interval' in mm
    rainFall = rainTick * 0.2794
    rainTick = 0
    
    # Print results
    print("\nTemperature: %0.1f C" % out_temp)
    print("Humidity: %0.1f %%" % humidity)
    print("Pressure: %0.2f hPa" % pressure)
    print("Case Temperature: %0.1f C" % case_temp) 
    print("Altitude = %0.2f meters" % bme.altitude)
    
    print( 'Wind Dir:    ' , windDir, ' (', val, ')')
    print ("Wind Speed: %0.2f km/h" % windSpeed)
    print ("Rainfall:   %0.2f  mm" % rainFall)
    print (" ")
    time.sleep(interval)


install.py
# installer for the weewx-bcrobo driver
# Copyright 2019 David Enstrom, all rights reserved
#
# See:    https://www.sparkfun.com/products/8942
#         https://www.bc-robotics.com/tutorials/raspberry-pi-weather-station-part-1/
#
# NOTE: We have moved to the 21st Century, so the database is in METRIC!
#       Therefore, if you have a database created already, then see:
#         http://weewx.com/docs/customizing.htm#Changing_the_unit_system

from setup import ExtensionInstaller

def loader():
    return BCRoboInstaller()

class BCRoboInstaller(ExtensionInstaller):
    def __init__(self):
        super(BCRoboInstaller, self).__init__(
            version="0.0.6",
            name='BCRobotics',
            description='Capture weather data from BC Robotics',
            author="David Enstrom",
            author_email="denstrom@rogers.com",
            config={
                'StdArchive': {
                    'record_generation': 'software'},
                'StdConvert': {
                    'target_unit': 'METRIC'},
                'StdQC': {
                    'MinMax': {
                        'barometer': '800, 1110, mbar',
                        'pressure': '800, 1110, mbar',
                        'outTemp': '-50, 49, degree_C',
                        'inTemp': '-50, 49, degree_C',
                        'windSpeed': '0, 300, km_per_hour'}},
                'StdReport': {
                    'StandardReport': {
                        'Units': {
                            'Groups': {
                                'group_altitude': 'meter',
                                'group_speed': 'km_per_hour',
                                'group_speed2': 'km_per_hour2',
                                'group_pressure': 'hPa',
                                'group_rain': 'mm',
                                'group_rainrate': 'mm_per_hour',
                                'group_temperature': 'degree_C',
                                'group_degree_day': 'degree_C',
                                'group_speed': 'km_per_hour'}}}}},

            files=[('bin/user', ['bin/user/BCRobotics.py'])]
        )
